### PR TITLE
SensESP's WebUI redesign and upgrade

### PR DESCRIPTION
**Pull request by JohnySeven**
For long time I wanted to redesign and upgrade web interface of SensESP devices to something more modern with few new features that could allow SensESP library users to add custom information into web interface with ease. This PR represents first wave of WebUI enhancements that features and addresses following:
1. Making WebUI efficient - this PR changes way how HTML web is servered by ESP32, all the resources are already compressed by GZIP compression. Flash requirements for WebUI has been reduced by 100 kB with adding better UI and more features!
2. Complete redesign of web interface using Twitter's Bootstrap framework v5.1 https://getbootstrap.com/docs/5.1/getting-started/introduction/, web interface uses only CSS components not Javascript components of Bootstrap. Note: we could eventually safe some space by reducing number of CSS components that are packed into Bootstrap css file.
3. The WebUI is SPA (single page application) type of web application so everything is driven by Javascript - nothing fancy here, just Pure Javascript. In my tests the WebUI loads from SH-ESP32 in about 500 ms.
4. WebUI now contains Status page (default), Configure page - shows drop down with configurations and user is able to configure sensor settings, Advanced hosts reset and restart commands.
4. I wanted to be able to add more information into WebUI, so I've introduced `UIOutput<T>` class that will allow inject any value into Status page of WebUI. All information stored in /info page in previous web interface version is now injected by using `UIOutput<T>` class. Here is sample of adding new information into status page:
`auto customUIOutput = new UIOutput<int>("UserInfo", 0);`

That's all for now - more is comming.

You can test WebUI here: http://webui.dytrych.cloud/  It's hosted on my server and shows all features of WebUI that are not in this PR like HTTPCommands and custom pages.

### CODE REVIEW

**Comment by JohnySeven** [comment]
https://user-images.githubusercontent.com/11419953/151420920-f0dbc0dd-aa80-4b71-b6b4-6ab65d414a99.mp4

Here is video of new WebUI in action from my phone.

**Comment by JohnySeven** [comment]
@ba58smith thanks Butch, I really appreciate it!

File: web/docroot/config/index.html [hunk 1]
```diff
@@ -1 +1 @@
-{"keys":["/gauge/curve","/gauge/inputs","/system/networking","/system/sk"]}
+{"keys":["/gauge/curve","/gauge/inputs","/system/networking","/system/sk"]}
```
Comment (mairas) [inline web/docroot/config/index.html:1]: Not sure if I understand any of the files in `web/docroot/config`. Accidental additions?

File: web/docroot/test [hunk 2]
```diff
@@ -0,0 +1,19 @@
+<div class="row g-3">
```
Comment (mairas) [inline web/docroot/test:1]: Accidental addition?

**Review by mairas — COMMENTED** [review]
Pretty great stuff. Just some minor stylistic bikeshed painting about variable naming and code formatting on my behalf.

I didn't really go through the JS files, though - I'm definitely not an expert on that front.

File: src/sensesp/net/networking.h [hunk 3]
```diff
@@ -62,6 +63,9 @@ class Networking : public Configurable,
   String preset_password = "";
   String preset_hostname = "";
   const char* wifi_manager_password_;
+  UILambdaOutput<String>* ssid_ = new UILambdaOutput<String>("SSID", [this]() { return this->ap_ssid; });
+  UIOutput<String>* mac_ = new UIOutput<String>("MAC", WiFi.macAddress());
+  UILambdaOutput<int8_t>* rssi_ = new UILambdaOutput<int8_t>("WiFi signal", [this]() { return WiFi.RSSI(); });
```
[the remark this answers is not in the document]
Comment (JohnySeven) [reply src/sensesp/net/networking.h:68]: Done

File: src/sensesp/net/web/css_bootstrap.h [hunk 4]
```diff
@@ -0,0 +1,2968 @@
+#include <pgmspace.h>
+const uint8_t PAGE_css_bootstrap[] PROGMEM = {
+		0x1f,0x8b,0x8,0x0,0xcb,0x9d,0xf2,0x61,
```
[the remark this answers is not in the document]
Comment (JohnySeven) [reply src/sensesp/net/web/css_bootstrap.h]: :-D I wasn't expecting somebody will!

File: src/sensesp/net/ws_client.cpp [hunk 5]
```diff
@@ -578,4 +577,22 @@ bool WSClient::set_configuration(const JsonObject& config) {
   return true;
 }
 
+String WSClient::get_connection_status()
```
[the remark this answers is not in the document]
Comment (JohnySeven) [reply src/sensesp/net/ws_client.cpp:580]: Done

File: src/sensesp/system/ui_output.cpp [hunk 6]
```diff
@@ -0,0 +1,10 @@
+#include "ui_output.h"
+namespace sensesp {
+  std::map<String, UIOutputBase*> uiOutputs;
```
[the remark this answers is not in the document]
Comment (JohnySeven) [reply src/sensesp/system/ui_output.cpp:3]: Done

File: src/sensesp/system/ui_output.h [hunk 7]
```diff
@@ -0,0 +1,75 @@
+#ifndef _SYSTEM_PROPERTY_H_
```
[the remark this answers is not in the document]
Comment (JohnySeven) [reply src/sensesp/system/ui_output.h:1]: Done

File: src/sensesp/system/ui_output.h [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+            }
+
+            virtual void set_json(const JsonObject& obj) { }
+     };
+
+     template<typename T>
+     class UILambdaOutput : public UIOutputBase
+     {
+         private:
+            std::function<T()> valueFunction_;
```
[the remark this answers is not in the document]
Comment (JohnySeven) [reply src/sensesp/system/ui_output.h:31]: Done

File: src/sensesp_base_app.cpp [hunk 9]
```diff
@@ -21,6 +21,9 @@ SensESPBaseApp::SensESPBaseApp() {
   // create the hostname_ observable - this needs to be done before
   // the builder methods are called
   hostname_ = new ObservableValue<String>(kDefaultHostname);
+  //Create buildInfo and system name properties
+  auto buildInfo = new UIOutput<String>("Build at", __DATE__ " " __TIME__);
```
[the remark this answers is not in the document]
Comment (JohnySeven) [reply src/sensesp_base_app.cpp:25]: I've added following 3 lines to be ready for adding SensESP version:
```
#ifdef SENSESP_VERSION
  auto verionInfo = new UIOutput<String>("SenseESP version", SENSESP_VERSION);
#endif
```

So we can use #define SENSESP_VERSION "v2.0" and update it manually or extras_script.py file automatically during build.

File: web/docroot/config/index.html (hunk 1 above)
Comment (JohnySeven) [reply web/docroot/config/index.html:1]: Those files are for WebUI testing using VS Code feature called "Live preview" that will start HTTPServer localy and you can edit WebUI and see how it works in browser.

File: web/docroot/test (hunk 2 above)
Comment (JohnySeven) [reply web/docroot/test:1]: Those files are for WebUI testing using VS Code feature called "Live preview" that will start HTTPServer localy and you can edit WebUI and see how it works in browser.

**Comment by JohnySeven** [comment]
@mairas all is done and it's ready to be merged.

File: web/docroot/config/index.html (hunk 1 above)
Comment (mairas) [reply web/docroot/config/index.html:1]: Roger that! It's all good, then.

File: src/sensesp_base_app.cpp (hunk 9 above)
[the remark this answers is not in the document]
Comment (mairas) [reply src/sensesp_base_app.cpp:25]: Let's do this slightly differently! I'm using `bump2version` to update hardcoded version strings in different configuration files. See `RELEASE_PROCESS.md` for details.

So, let's create a new header file, say, `src/sensesp/sensesp_version.h` which has the following contents:

```
#ifndef __SRC_SENSESP_SENSESP_VERSION_H_
#define __SRC_SENSESP_SENSESP_VERSION_H_ 

const char* kSensESPVersion = "2.0.1-alpha";

#endif
```

That file can then be included in `sensesp_base_app.cpp`.

Then, you can declare

```
auto version_info = new UIOutput<String>("SenseESP version", kSensESPVersion);
```

without the ifdefs.

Then, `.bumpversion.cfg` can be modified to update `sensesp_version.h` whenever the version number is bumped.

If you prefer, I can also make these changes in this PR once everything else is done.

File: web/docroot/test (hunk 2 above)
Comment (mairas) [reply web/docroot/test:1]: Roger!

**Comment by mairas** [comment]
There're a couple of test builds failing. I suppose you missed renaming a declaration somewhere.

https://github.com/SignalK/SensESP/runs/4997636871?check_suite_focus=true#step:5:273

I like to do all renaming myself using the VSCode Rename Symbol command (F2). It's quite good in picking up the related declarations and all uses, so there's less to go wrong if you're doing some major refactoring.

**Comment by mairas** [comment]
Other than that (and the SensESP version UIOutput change), looks good to go!

File: src/sensesp/net/http_server.cpp [hunk 10]
```diff
@@ -207,24 +217,26 @@ void HTTPServer::handle_device_restart(AsyncWebServerRequest* request) {
 }
 
 void HTTPServer::handle_info(AsyncWebServerRequest* request) {
-  auto* response = request->beginResponseStream("text/plain");
-
+  auto* response = request->beginResponseStream("application/json");
+  char buff[128];
```
Comment (mairas) [inline src/sensesp/net/http_server.cpp:221]: I think this guy actually gave a compilation warning. Not used?

File: src/sensesp/system/ui_output.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      : UIOutputBase(name) {
+    value_function_ = function;
+  }
+
+  T get() { return value_function_(); }
+
+  void set_json(const JsonObject& obj) override { obj[name_] = get(); }
+};
+
+extern std::map<String, UIOutputBase*> uiOutputs;
```
Comment (mairas) [inline src/sensesp/system/ui_output.h:42]: I think this is the culprit?

File: src/sensesp/net/http_server.cpp (hunk 10 above)
Comment (JohnySeven) [reply src/sensesp/net/http_server.cpp:221]: Yes :-) Removed, thanks!

File: src/sensesp/system/ui_output.h (hunk 11 above)
Comment (JohnySeven) [reply src/sensesp/system/ui_output.h:42]: Correct, it's fixed!

**Comment by JohnySeven** [comment]
Ok, now it's really ready! All checks have passed, so whenever you have time @mairas !

**Comment by mairas** [comment]
I did some small finishing touches. I'll merge as soon as the builds pass.

**Pull request merged**